graphs/index.md

```diff
@@ -0,0 +1,706 @@
+quickly for evaluation the scope of this proposal was limited to a subset
+of these requirements. In particular, the serialization functionality (9),
+backend interoperability (10), and a profiling/debugging interface (3) were
+omitted. As these are not easy to abstract over a number of backends without
+significant investigation. It is also hoped these features can be exposed as
+additive changes to the API, and so in introduced in future versions of the
+extension.
+
+Another reason for deferring a serialize/deserialize API (9) is that its scope
+could extend from emitting the graph in a binary format, to emitting a
+standardized IR format that enables further device specific graph optimizations.
+
+As described in [Design Discussion](#no-explicit-graph-api) this proposal
+defines an implicit recording design for constructing the graph rather than an
+explicit graph construction API. It is envisioned that an explicit API could
+better express several of the requirements above. In particular the ability
+for a user to specify exactly what device they'd like a node to run on when
+constructing multiple device graphs (7). See the section on
+[single queue submit](#single-queue-submit) for more discussion on the multiple
+device use case.
+
+## Extension
+
+### SYCL Graph Definition
+
+A SYCL graph is a collection of nodes and edges. From the SYCL perspective, this
+graph will be acyclic and directed (DAG) as users cannot express a cycle in the API.
+
+#### Node
+
+Nodes in a SYCL graph are defined as each of the command group submissions of the
+program. Each submission encompasses either one or both of a.) some data movement,
+b.) a single asynchronous kernel launch. Nodes cannot define forward edges, only
+backwards (i.e. kernels can only create dependencies on things that have already
+happened). This means that transparently a node can depend on a previously
+recorded graph (sub-graph), which works by creating edges to the individual nodes
+in the old graph. Explicit memory operations without kernels, such as a memory
+copy, are still classed as nodes under this definition, as the [SYCL 2020
+specification states][explicit-memory-ops] that these can be seen as specialized
+kernels executing on the device.
+
+#### Edge
+
+An edge in the SYCL-graph represents a data dependency between two nodes. These
+dependencies are expressed by the user code through buffer accessors. There is
+also the partial ability to track USM data dependencies provided the pointers
+used in the graph nodes are the same. With the limitation that a node taking
+an offset USM pointer input will not be identified as having an edge to
+another node taking a pointer input to the base address of the same USM
+allocation.
+
+### API Modifications
+
+```cpp
+namespace sycl {
+namespace ext::codeplay {
+
+// State of a queue, returned by info::queue::state
+enum class queue_state {
+  executing,
+  recording
+};
+
+// State of a graph
+enum class graph_state {
+  recordable,
```

Comment:
Naming this state `modifiable` or `editable` might be more generic, and better support future extensions (such as an Explicit API). Or would you consider adding an extra state?

Reply:
We weren't considering adding extra state, so calling `recordable` more generic would indeed to better introducing an explicit API. My worry with `modifiable` or `editable` is that the user can perform a kind of `modify` or `edit`  of the graph when we update it, either through whole-graph update or individual node update in an Explicit API.

Would probably tend towards `constructable`, `buildable`, `composable`, or something along those lines. Will have a think.

Reply:
Decided to make this `modifiable` for the moment to reduce the diff with the oneAPI extension, once they are merged we can always revisit this.